### ISSUE TITLE
`hipMemcpy3D` is broken

### ISSUE THREAD

**Issue by psychocoderHPC** [ROCm-Developer-Tools/HIP#1790]
`hipMemcpy3D` is working incorrectly. When I copy a array's with different pitch the destination offset is wrong.

The example is slicing a 4x4x4 integer *C pointer array*
  - the offset is 3x0x0 and applied directly to the pointer addresses
  - the extent is 1x3x3
The destination array is initialized with the value `133` and the values in the source array are enumerated.

```C++
#include <stdio.h>
#include <hip_runtime.h>


__global__ void Fill(int* ptr, size_t pitch, const int numElements)
{
    for(int z = 0; z < numElements; ++z)
    {
        for(int y = 0; y < numElements; ++y)
        {
            for(int x = 0; x < numElements; ++x)
            {
                int idx = z * numElements * numElements + y * numElements + x;
                ((int*)((char*)ptr + z * numElements * pitch + y * pitch))[x] = idx;
            }
        }
    }
}

int main(int argc, char *argv[])
{

    hipPitchedPtr dPtr;

    auto extent = make_hipExtent(4 * sizeof(int),4,4);

    hipMalloc3D(&dPtr, extent);

    int* hPtr;
    hipMallocHost((void**)&hPtr, sizeof(int)*4*4*4);

    for(int i = 0; i < 4 * 4 * 4; ++i)
    {
        hPtr[i] = 133;
    }


    hipLaunchKernelGGL(HIP_KERNEL_NAME(Fill),1,1,0,0,(int*)dPtr.ptr,dPtr.pitch, 4);

    auto srcColumn = make_hipPitchedPtr(
        (char*)dPtr.ptr + 3 * sizeof(int),
        dPtr.pitch,
        4,
        4
    );

    auto dstColumn = make_hipPitchedPtr(
        hPtr + 3,
        4 * sizeof(int),
        4,
        4
    );

    auto copyExtent = make_hipExtent(1*sizeof(int),3,3);

    hipMemcpy3DParms window;
    window.srcArray = nullptr;
    window.dstArray = nullptr;

    window.srcPtr = srcColumn;
    window.srcPos = make_hipPos(0,0,0);
    window.dstPtr = dstColumn;
    window.dstPos = make_hipPos(0,0,0);
    window.extent = copyExtent;
    window.kind = hipMemcpyDeviceToHost;
    
    hipMemcpy3D(
        &window
    );

    hipError_t err = hipGetLastError();
    if( hipSuccess != err) 
       printf("error: %s\n", hipGetErrorString( err));

    for(int z = 0; z < 4; ++z)
    {
        for(int y = 0; y < 4; ++y)
        {
            for(int x = 0; x < 4; ++x)
            {
                int idx = z * 4 * 4 + y * 4 + x;

                printf("%i ", hPtr[idx]);
            }
            printf("\n");
        }
        printf("\n");
        printf("\n");
    }
}
```

output is:
```
133 133 133 3 
133 133 133 7 
133 133 133 11 
133 133 133 19 


133 133 133 23 
133 133 133 27 
133 133 133 35 
133 133 133 39 


133 133 133 43 
133 133 133 133 
133 133 133 133 
133 133 133 133 


133 133 133 133 
133 133 133 133 
133 133 133 133 
133 133 133 133
```
but should be

```
133 133 133 3 
133 133 133 7 
133 133 133 11 
133 133 133 133 


133 133 133 19 
133 133 133 23 
133 133 133 27 
133 133 133 133 


133 133 133 35 
133 133 133 39 
133 133 133 43 
133 133 133 133 


133 133 133 133 
133 133 133 133 
133 133 133 133 
133 133 133 133
```

The bug is a wrong offset in the destination array.
https://github.com/ROCm-Developer-Tools/HIP/blob/753277422a0637ff2bf541c7b43dfaaa2b12c9f6/src/hip_memory.cpp#L1673-L1674
is must be
```C++
unsigned char* dst = (unsigned char*)dstPtr + i * p->dstPtr.ysize * dstPitch + j * dstPitch;
```

**Comment by psychocoderHPC** [comment]
I do not provided the fix as PR because #1791 should be fixed too.

**Comment by psychocoderHPC** [comment]
@gargrahul is working on a fix in #1798
